File: ios/Tables/TableCellPositionMap.swift
import Foundation

struct TableCellPositionMap {
    struct Binding: Equatable {
        let cellSourcePosition: UInt32
        let documentRevision: UInt64
        let positionEpoch: UInt64
    }

    struct Segment: Equatable {
        let localScalarRange: Range<UInt32>
        let globalScalarStart: UInt32
    }

    let binding: Binding
    let segments: [Segment]

    init(binding: Binding, segments: [Segment]) {
        self.binding = binding
        self.segments = segments.sorted { $0.localScalarRange.lowerBound < $1.localScalarRange.lowerBound }
    }

    func globalScalar(
        forLocalScalar localScalar: UInt32,
        currentRevision: UInt64? = nil,
        currentEpoch: UInt64? = nil
    ) -> UInt32? {
        if let currentRevision, currentRevision != binding.documentRevision { return nil }
        if let currentEpoch, currentEpoch != binding.positionEpoch { return nil }
        var resolved: UInt32?
        for segment in segments where segment.localScalarRange.contains(localScalar) {
            let offset = localScalar - segment.localScalarRange.lowerBound
            let (candidate, overflow) = segment.globalScalarStart.addingReportingOverflow(offset)
            guard !overflow, resolved == nil || resolved == candidate else { return nil }
            resolved = candidate
        }
        return resolved
    }

    func localScalar(
        forGlobalScalar globalScalar: UInt32,
        currentRevision: UInt64? = nil,
        currentEpoch: UInt64? = nil
    ) -> UInt32? {
        if let currentRevision, currentRevision != binding.documentRevision { return nil }
        if let currentEpoch, currentEpoch != binding.positionEpoch { return nil }
        var resolved: UInt32?
        for segment in segments where globalScalar >= segment.globalScalarStart {
            let width = segment.localScalarRange.upperBound - segment.localScalarRange.lowerBound
            let offset = globalScalar - segment.globalScalarStart
            guard offset < width else { continue }
            let candidate = segment.localScalarRange.lowerBound + offset
            guard resolved == nil || resolved == candidate else { return nil }
            resolved = candidate
        }
        return resolved
    }

    func globalScalarRange(fromLocalScalar start: UInt32, toLocalScalar end: UInt32) -> (from: UInt32, to: UInt32)? {
        guard start <= end,
              let globalStart = globalScalar(forLocalScalar: start),
              let globalEnd = globalScalar(forLocalScalar: end),
              globalEnd >= globalStart,
              UInt64(globalEnd) - UInt64(globalStart) == UInt64(end) - UInt64(start)
        else { return nil }
        let limit = UInt64(end) + 1
        var covered = UInt64(start)
        for segment in segments {
            let lower = max(UInt64(segment.localScalarRange.lowerBound), UInt64(start))
            let upper = min(UInt64(segment.localScalarRange.upperBound), limit)
            guard lower < upper else { continue }
            guard lower <= covered else { return nil }
            let actual = UInt64(segment.globalScalarStart) + lower - UInt64(segment.localScalarRange.lowerBound)
            let expected = UInt64(globalStart) + lower - UInt64(start)
            guard actual == expected else { return nil }
            covered = max(covered, upper)
        }
        guard covered == limit else { return nil }
        return (globalStart, globalEnd)
    }

    func globalScalar(forLocalUTF16 offset: Int, in text: String) -> UInt32? {
        globalScalar(forLocalScalar: PositionBridge.utf16OffsetToScalar(offset, in: text))
    }

    func globalScalarRange(forLocalUTF16 range: NSRange, in text: String) -> (UInt32, UInt32)? {
        guard range.location != NSNotFound,
              range.location >= 0,
              range.length >= 0,
              range.location <= text.utf16.count - range.length
        else { return nil }
        return globalScalarRange(
            fromLocalScalar: PositionBridge.utf16OffsetToScalar(range.location, in: text),
            toLocalScalar: PositionBridge.utf16OffsetToScalar(range.location + range.length, in: text)
        )
    }
}

File: android/src/main/java/com/apollohg/editor/tables/TableCellPositionMap.kt
package com.apollohg.editor.tables

import com.apollohg.editor.PositionBridge

internal class TableCellPositionMap(
    val binding: Binding,
    segments: List<Segment>
) {
    data class Binding(
        val cellSourcePos: Long,
        val revision: String,
        val epoch: String
    )

    data class Segment(
        val localScalarStart: Int,
        val localScalarEndExclusive: Int,
        val globalScalarStart: Int
    )

    data class ScalarRange(val from: Int, val to: Int)

    private val segments = segments.sortedBy { it.localScalarStart }

    fun globalScalarForLocalScalar(
        localScalar: Int,
        currentRevision: String? = null,
        currentEpoch: String? = null
    ): Int? {
        if (!matchesCurrent(currentRevision, currentEpoch)) return null
        var resolved: Int? = null
        for (segment in segments) {
            if (localScalar < segment.localScalarStart || localScalar >= segment.localScalarEndExclusive) continue
            val candidate = add(
                segment.globalScalarStart.toLong(),
                localScalar.toLong() - segment.localScalarStart.toLong()
            ) ?: return null
            if (resolved != null && resolved != candidate) return null
            resolved = candidate
        }
        return resolved
    }

    fun localScalarForGlobalScalar(
        globalScalar: Int,
        currentRevision: String? = null,
        currentEpoch: String? = null
    ): Int? {
        if (!matchesCurrent(currentRevision, currentEpoch)) return null
        var resolved: Int? = null
        for (segment in segments) {
            val width = segment.localScalarEndExclusive.toLong() - segment.localScalarStart.toLong()
            val offset = globalScalar.toLong() - segment.globalScalarStart.toLong()
            if (width <= 0 || offset < 0 || offset >= width) continue
            val candidate = add(segment.localScalarStart.toLong(), offset) ?: return null
            if (resolved != null && resolved != candidate) return null
            resolved = candidate
        }
        return resolved
    }

    fun globalScalarRange(fromLocalScalar: Int, toLocalScalar: Int): ScalarRange? {
        if (fromLocalScalar > toLocalScalar) return null
        val globalStart = globalScalarForLocalScalar(fromLocalScalar) ?: return null
        val globalEnd = globalScalarForLocalScalar(toLocalScalar) ?: return null
        if (globalEnd.toLong() - globalStart.toLong() != toLocalScalar.toLong() - fromLocalScalar.toLong()) return null

        val rangeEndExclusive = toLocalScalar.toLong() + 1L
        var covered = fromLocalScalar.toLong()
        for (segment in segments) {
            val lower = maxOf(segment.localScalarStart.toLong(), fromLocalScalar.toLong())
            val upper = minOf(segment.localScalarEndExclusive.toLong(), rangeEndExclusive)
            if (lower >= upper) continue
            if (lower > covered) return null
            val actual = segment.globalScalarStart.toLong() + lower - segment.localScalarStart.toLong()
            val expected = globalStart.toLong() + lower - fromLocalScalar.toLong()
            if (actual != expected) return null
            covered = maxOf(covered, upper)
        }
        return if (covered == rangeEndExclusive) ScalarRange(globalStart, globalEnd) else null
    }

    fun globalScalarForLocalUtf16(offset: Int, text: String): Int? {
        if (offset !in 0..text.length) return null
        return globalScalarForLocalScalar(PositionBridge.utf16ToScalar(offset, text))
    }

    fun globalScalarRangeForLocalUtf16(start: Int, end: Int, text: String): ScalarRange? {
        if (start < 0 || end < start || end > text.length) return null
        return globalScalarRange(
            PositionBridge.utf16ToScalar(start, text),
            PositionBridge.utf16ToScalar(end, text)
        )
    }

    fun isCurrent(currentRevision: String, currentEpoch: String): Boolean =
        binding.revision == currentRevision && binding.epoch == currentEpoch

    fun hasValidSegments(): Boolean = segments.isNotEmpty() && segments.all { segment ->
        segment.localScalarStart >= 0 &&
            segment.localScalarEndExclusive > segment.localScalarStart &&
            segment.globalScalarStart >= 0 &&
            hasRepresentableGlobalEnd(segment)
    }

    private fun matchesCurrent(currentRevision: String?, currentEpoch: String?): Boolean =
        (currentRevision == null || binding.revision == currentRevision) &&
            (currentEpoch == null || binding.epoch == currentEpoch)

    private fun add(start: Long, offset: Long): Int? {
        val value = start + offset
        return value.takeIf { it in Int.MIN_VALUE.toLong()..Int.MAX_VALUE.toLong() }?.toInt()
    }

    private fun hasRepresentableGlobalEnd(segment: Segment): Boolean =
        segment.globalScalarStart.toLong() +
            (segment.localScalarEndExclusive.toLong() - segment.localScalarStart.toLong() - 1L) <=
            Int.MAX_VALUE.toLong()
}
